File: kalitech-engine/src/main/resources/kalitech/engine/modules/Terrain/Terrain.js
// FILE: resources/kalitech/builtin/Terrain.js
// Author: Calista Verner (v2 API overhaul)
"use strict";

const META = Object.freeze({
    moduleId: "terrain",
    globalName: "TERR",
    version: "2.0.0",
    engineMin: "0.1.0",
    description: "Declarative terrain builder (plane/quad/heightmap/heights/noise) + physics + edit/query",
});

const {isObj, num, i32} = require("./helpers/TerrainTypes.js");
const {TerrainHeights} = require("./helpers/TerrainHeights.js");
const {TerrainPhysics} = require("./helpers/TerrainPhysics.js");
const {TerrainCreateHelper} = require("./helpers/TerrainCreateHelper.js");

function makeApi(engine) {
    if (!engine) throw new Error("[TERR] engine is required");
    const terr = (engine.terrain && typeof engine.terrain === "function") ? engine.terrain() : null;
    if (!terr) throw new Error("[TERR] engine.terrain() is not available");

    const heightsApi = new TerrainHeights(terr);
    const physicsApi = new TerrainPhysics(engine);

    function terrain(cfg) {
        const c = isObj(cfg) ? Object.assign({}, cfg) : {};
        const physCfg = c.physics;
        if (physCfg != null) delete c.physics;

        if (c.size || c.patchSize) TerrainHeights.validateTerrainDims(i32(c.size, 0), i32(c.patchSize, 0));

        const surface = terr.terrain(c);
        return physicsApi.withBody(terr, surface, physCfg, "mesh");
    }

    function terrainHeights(cfg) {
        const c = isObj(cfg) ? Object.assign({}, cfg) : {};
        const physCfg = c.physics;
        if (physCfg != null) delete c.physics;

        const heights = c.heights;
        if (heights != null) c.heights = TerrainHeights.toFloat32Array(heights);

        const size = i32(c.size, 0) || TerrainHeights.inferSizeFromHeights(c.heights);
        if (size > 0) c.size = size;

        if (c.size || c.patchSize) TerrainHeights.validateTerrainDims(i32(c.size, 0), i32(c.patchSize, 0));
        TerrainHeights.assertHeightsMatchSize(c.heights, c.size, "terrainHeights");

        const surface = terr.terrainHeights(c);
        return physicsApi.withBody(terr, surface, physCfg, "mesh");
    }

    function quad(cfg) {
        const c = isObj(cfg) ? Object.assign({}, cfg) : {};
        const physCfg = c.physics;
        if (physCfg != null) delete c.physics;

        const surface = terr.quad(c);
        return physicsApi.withBody(terr, surface, physCfg, "mesh");
    }

    function plane(cfg) {
        const c = isObj(cfg) ? Object.assign({}, cfg) : {};
        const physCfg = c.physics;
        if (physCfg != null) delete c.physics;

        const surface = terr.plane(c);
        return physicsApi.withBody(terr, surface, physCfg, "mesh");
    }

    function physics(surfaceHandleOrId, cfg) {
        if (!surfaceHandleOrId) throw new Error("TERR.physics(surface,cfg): surface handle/id required");
        return physicsApi.withBody(terr, surfaceHandleOrId, cfg || {}, "mesh");
    }

    function material(surfaceHandle, materialHandleOrCfg) {
        return terr.material(surfaceHandle, materialHandleOrCfg);
    }

    function uv(surfaceHandle, cfg) {
        return terr.uv(surfaceHandle, cfg);
    }

    function lod(surfaceHandle, cfg) {
        return terr.lod(surfaceHandle, cfg || {});
    }

    function scale(surfaceHandle, xzScale, cfg) {
        return terr.scale(surfaceHandle, num(xzScale, 1.0), cfg || null);
    }

    function heightAt(surfaceHandle, x, z, world) {
        if (world === undefined) return terr.heightAt(surfaceHandle, num(x, 0), num(z, 0));
        return terr.heightAt(surfaceHandle, num(x, 0), num(z, 0), !!world);
    }

    function normalAt(surfaceHandle, x, z, world) {
        if (world === undefined) return terr.normalAt(surfaceHandle, num(x, 0), num(z, 0));
        return terr.normalAt(surfaceHandle, num(x, 0), num(z, 0), !!world);
    }

    function setHeightmap(surfaceHandle, heights, size, rebuild) {
        if (isObj(heights)) return terr.setHeightmap(surfaceHandle, heights);

        const h = (heights instanceof Float32Array) ? heights : TerrainHeights.toFloat32Array(heights);
        const s = (size | 0) || TerrainHeights.inferSizeFromHeights(h);

        if (s > 0) TerrainHeights.assertHeightsMatchSize(h, s, "setHeightmap");

        return terr.setHeightmap(surfaceHandle, {
            heights: h,
            size: s || undefined,
            rebuild: (rebuild === undefined) ? true : !!rebuild,
        });
    }

    function heightmap(surfaceHandle) {
        return TerrainHeights.toFloat32Array(terr.heightmap(surfaceHandle));
    }

    function setHeight(surfaceHandle, x, z, height, world) {
        if (world === undefined) return terr.setHeight(surfaceHandle, num(x, 0), num(z, 0), num(height, 0));
        return terr.setHeight(surfaceHandle, num(x, 0), num(z, 0), num(height, 0), !!world);
    }

    function adjustHeight(surfaceHandle, x, z, delta, world) {
        if (world === undefined) return terr.adjustHeight(surfaceHandle, num(x, 0), num(z, 0), num(delta, 0));
        return terr.adjustHeight(surfaceHandle, num(x, 0), num(z, 0), num(delta, 0), !!world);
    }

    function rebuild(surfaceHandle) {
        return terr.rebuild(surfaceHandle);
    }

    function attach(surfaceHandle, entityId) {
        return terr.attach(surfaceHandle, entityId | 0);
    }

    function detach(surfaceHandle) {
        return terr.detach(surfaceHandle);
    }

    const heightsNS = Object.freeze({
        perlin: (cfg) => heightsApi.perlin(cfg || {}),
        ridged: (cfg) => heightsApi.ridged(cfg || {}),
        sizeOf: TerrainHeights.inferSizeFromHeights,
        toF32: TerrainHeights.toFloat32Array,
    });

    const createHelper = new TerrainCreateHelper(
        engine,
        terr,
        heightsApi,
        physicsApi,
        {
            terrain,
            terrainHeights,
            quad,
            plane,
            physics,
            material,
            uv,
            lod,
            scale,
            heightAt,
            normalAt,
            setHeightmap,
            heightmap,
            setHeight,
            adjustHeight,
            rebuild,
            attach,
            detach
        }
    );

    function create(cfg) {
        return createHelper.create(cfg);
    }

    return Object.freeze({
        META,
        create,
        heights: heightsNS,

        terrain,
        terrainHeights,
        quad,
        plane,

        physics,
        material,
        uv,
        lod,
        scale,
        heightAt,
        normalAt,

        setHeightmap,
        heightmap,
        setHeight,
        adjustHeight,
        rebuild,

        attach,
        detach,
    });
}

module.exports = function TerrainModule(engine, K) {
    return makeApi(engine, K);
};
module.exports.META = META;

File: modulesSrc/TerrainModule/src/main/resources/kalitech/engine/modules/Terrain/Terrain.lua
local luaRuntime = require("@builtin/lua_runtime")
local Error = luaRuntime.Error
local LuaConstruct = luaRuntime.LuaConstruct
local LuaTableMerge = luaRuntime.LuaTableMerge
local LuaTableRemove = luaRuntime.LuaTableRemove
META = KObject:freeze({
    moduleId = "terrain",
    version = "2.0.0",
    engineMin = "0.1.0",
    description = "Declarative terrain builder (plane/quad/heightmap/heights/noise) + physics + edit/query"
})
local lua_require_result_0 = require("./helpers/TerrainTypes.lua")
isObj = lua_require_result_0.isObj
num = lua_require_result_0.num
i32 = lua_require_result_0.i32
local lua_require_result_1 = require("./helpers/TerrainHeights.lua")
TerrainHeights = lua_require_result_1.TerrainHeights
local lua_require_result_2 = require("./helpers/TerrainPhysics.lua")
TerrainPhysics = lua_require_result_2.TerrainPhysics
local lua_require_result_3 = require("./helpers/TerrainCreateHelper.lua")
TerrainCreateHelper = lua_require_result_3.TerrainCreateHelper
function makeApi(self, engine)
    if not engine then
        error(
            LuaConstruct(Error, "[TERR] engine is required"),
            0
        )
    end
    local lua_temp_4
    if engine.terrain and KTypeOf(engine.terrain) == "function" then
        lua_temp_4 = engine:terrain()
    else
        lua_temp_4 = nil
    end
    local terr = lua_temp_4
    if not terr then
        error(
            LuaConstruct(Error, "[TERR] engine.terrain() is not available"),
            0
        )
    end
    local heightsApi = LuaConstruct(TerrainHeights, terr)
    local physicsApi = LuaConstruct(TerrainPhysics, engine)
    local function terrain(self, cfg)
        local lua_isObj_result_5
        if isObj(_G, cfg) then
            lua_isObj_result_5 = LuaTableMerge({}, cfg)
        else
            lua_isObj_result_5 = {}
        end
        local c = lua_isObj_result_5
        local physCfg = c.physics
        if physCfg ~= nil then
            LuaTableRemove(c, "physics")
        end
        if c.size or c.patchSize then
            TerrainHeights:validateTerrainDims(
                i32(_G, c.size, 0),
                i32(_G, c.patchSize, 0)
            )
        end
        local surface = terr:terrain(c)
        return physicsApi:withBody(terr, surface, physCfg, "mesh")
    end
    local function terrainHeights(self, cfg)
        local lua_isObj_result_6
        if isObj(_G, cfg) then
            lua_isObj_result_6 = LuaTableMerge({}, cfg)
        else
            lua_isObj_result_6 = {}
        end
        local c = lua_isObj_result_6
        local physCfg = c.physics
        if physCfg ~= nil then
            LuaTableRemove(c, "physics")
        end
        local heights = c.heights
        if heights ~= nil then
            c.heights = TerrainHeights:toFloatArray(heights)
        end
        local size = i32(_G, c.size, 0) or TerrainHeights:inferSizeFromHeights(c.heights)
        if size > 0 then
            c.size = size
        end
        if c.size or c.patchSize then
            TerrainHeights:validateTerrainDims(
                i32(_G, c.size, 0),
                i32(_G, c.patchSize, 0)
            )
        end
        TerrainHeights:assertHeightsMatchSize(c.heights, c.size, "terrainHeights")
        local surface = terr:terrainHeights(c)
        return physicsApi:withBody(terr, surface, physCfg, "mesh")
    end
    local function quad(self, cfg)
        local lua_isObj_result_7
        if isObj(_G, cfg) then
            lua_isObj_result_7 = LuaTableMerge({}, cfg)
        else
            lua_isObj_result_7 = {}
        end
        local c = lua_isObj_result_7
        local physCfg = c.physics
        if physCfg ~= nil then
            LuaTableRemove(c, "physics")
        end
        local surface = terr:quad(c)
        return physicsApi:withBody(terr, surface, physCfg, "mesh")
    end
    local function plane(self, cfg)
        local lua_isObj_result_8
        if isObj(_G, cfg) then
            lua_isObj_result_8 = LuaTableMerge({}, cfg)
        else
            lua_isObj_result_8 = {}
        end
        local c = lua_isObj_result_8
        local physCfg = c.physics
        if physCfg ~= nil then
            LuaTableRemove(c, "physics")
        end
        local surface = terr:plane(c)
        return physicsApi:withBody(terr, surface, physCfg, "mesh")
    end
    local function physics(self, surfaceHandleOrId, cfg)
        if not surfaceHandleOrId then
            error(
                LuaConstruct(Error, "TERR.physics(surface,cfg): surface handle/id required"),
                0
            )
        end
        return physicsApi:withBody(terr, surfaceHandleOrId, cfg or ({}), "mesh")
    end
    local function material(self, surfaceHandle, materialHandleOrCfg)
        return terr:material(surfaceHandle, materialHandleOrCfg)
    end
    local function uv(self, surfaceHandle, cfg)
        return terr:uv(surfaceHandle, cfg)
    end
    local function lod(self, surfaceHandle, cfg)
        return terr:lod(surfaceHandle, cfg or ({}))
    end
    local function scale(self, surfaceHandle, xzScale, cfg)
        return terr:scale(
            surfaceHandle,
            num(_G, xzScale, 1),
            cfg or nil
        )
    end
    local function heightAt(self, surfaceHandle, x, z, world)
        if world == nil then
            return terr:heightAt(
                surfaceHandle,
                num(_G, x, 0),
                num(_G, z, 0)
            )
        end
        return terr:heightAt(
            surfaceHandle,
            num(_G, x, 0),
            num(_G, z, 0),
            not not world
        )
    end
    local function normalAt(self, surfaceHandle, x, z, world)
        if world == nil then
            return terr:normalAt(
                surfaceHandle,
                num(_G, x, 0),
                num(_G, z, 0)
            )
        end
        return terr:normalAt(
            surfaceHandle,
            num(_G, x, 0),
            num(_G, z, 0),
            not not world
        )
    end
    local function setHeightmap(self, surfaceHandle, heights, size, rebuild)
        if isObj(_G, heights) then
            return terr:setHeightmap(surfaceHandle, heights)
        end
        local h = TerrainHeights:toFloatArray(heights)
        local s = bit32.bor(size, 0) or TerrainHeights:inferSizeFromHeights(h)
        if s > 0 then
            TerrainHeights:assertHeightsMatchSize(h, s, "setHeightmap")
        end
        local lua_terr_setHeightmap_13 = terr.setHeightmap
        local lua_surfaceHandle_12 = surfaceHandle
        local lua_temp_11 = s or nil
        local lua_temp_10
        if rebuild == nil then
            lua_temp_10 = true
        else
            lua_temp_10 = not not rebuild
        end
        return lua_terr_setHeightmap_13(terr, lua_surfaceHandle_12, {heights = h, size = lua_temp_11, rebuild = lua_temp_10})
    end
    local function heightmap(self, surfaceHandle)
        return TerrainHeights:toFloatArray(terr:heightmap(surfaceHandle))
    end
    local function setHeight(self, surfaceHandle, x, z, height, world)
        if world == nil then
            return terr:setHeight(
                surfaceHandle,
                num(_G, x, 0),
                num(_G, z, 0),
                num(_G, height, 0)
            )
        end
        return terr:setHeight(
            surfaceHandle,
            num(_G, x, 0),
            num(_G, z, 0),
            num(_G, height, 0),
            not not world
        )
    end
    local function adjustHeight(self, surfaceHandle, x, z, delta, world)
        if world == nil then
            return terr:adjustHeight(
                surfaceHandle,
                num(_G, x, 0),
                num(_G, z, 0),
                num(_G, delta, 0)
            )
        end
        return terr:adjustHeight(
            surfaceHandle,
            num(_G, x, 0),
            num(_G, z, 0),
            num(_G, delta, 0),
            not not world
        )
    end
    local function rebuild(self, surfaceHandle)
        return terr:rebuild(surfaceHandle)
    end
    local function attachEntity(self, surfaceHandle, entityUuid)
        return terr:attachEntity(surfaceHandle, entityUuid)
    end
    local function attach(self, surfaceHandle, entityUuid)
        return attachEntity(_G, surfaceHandle, entityUuid)
    end
    local function detach(self, surfaceHandle)
        return terr:detach(surfaceHandle)
    end
    local heightsNS = KObject:freeze({
        perlin = function(lua_, cfg) return heightsApi:perlin(cfg or ({})) end,
        ridged = function(lua_, cfg) return heightsApi:ridged(cfg or ({})) end,
        sizeOf = TerrainHeights.inferSizeFromHeights,
        toArray = TerrainHeights.toFloatArray
    })
    local createHelper = LuaConstruct(
        TerrainCreateHelper,
        engine,
        terr,
        heightsApi,
        physicsApi,
        {
            terrain = terrain,
            terrainHeights = terrainHeights,
            quad = quad,
            plane = plane,
            physics = physics,
            material = material,
            uv = uv,
            lod = lod,
            scale = scale,
            heightAt = heightAt,
            normalAt = normalAt,
            setHeightmap = setHeightmap,
            heightmap = heightmap,
            setHeight = setHeight,
            adjustHeight = adjustHeight,
            rebuild = rebuild,
            attach = attach,
            attachEntity = attachEntity,
            detach = detach
        }
    )
    local function create(self, cfg)
        return createHelper:create(cfg)
    end
    return KObject:freeze({
        META = META,
        create = create,
        heights = heightsNS,
        terrain = terrain,
        terrainHeights = terrainHeights,
        quad = quad,
        plane = plane,
        physics = physics,
        material = material,
        uv = uv,
        lod = lod,
        scale = scale,
        heightAt = heightAt,
        normalAt = normalAt,
        setHeightmap = setHeightmap,
        heightmap = heightmap,
        setHeight = setHeight,
        adjustHeight = adjustHeight,
        rebuild = rebuild,
        attach = attach,
        attachEntity = attachEntity,
        detach = detach
    })
end
local moduleFactory = function(self, engine, K)
    return makeApi(_G, engine, K)
end
return setmetatable({META = META}, {
    __call = function(_, ...)
        return moduleFactory(...)
    end
})
